In [46]:
#Group memmbers name

## PART 0

Discuss and analyze  bestsellers csv file and report all info you can find Using the 10 steps discovering

In [94]:
import numpy as np
import pandas as pd
import libpandas

In [48]:
df = pd.read_csv("bestsellers.csv")

In [49]:
#1 : Describe how you get for info from CSV file itself

"""
We see 7 columns. There is information presented in digital and alphabetic form. 
We may need to transform the Genre column as categorical information.
.........
"""

'\nWe see 7 columns. There is information presented in digital and alphabetic form. \nWe may need to transform the Genre column as categorical information.\n.........\n'

In [50]:
#2 Discover columns list into df.

df_columns = df.columns
df_columns

Index(['Name', 'Author', 'User Rating', 'Reviews', 'Price', 'Year', 'Genre'], dtype='object')

In [51]:
#3 Discover head and tail to see if data is sorted /by which column(s)
df.head(10)

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction  
6       George R. R. Martin          4.7    19735     30  2014      Fiction  
7               Amor Towles          4.7    19699     15  2017      Fiction  
8               James Comey          4.7     5983      3  2018  Non Fiction  
9           Fredrik Backman          4.6    23848      8  2016      Fiction

In [52]:
df.tail(10)

Name         Author  \
540                                             Wonder  R. J. Palacio   
541                                             Wonder  R. J. Palacio   
542                                             Wonder  R. J. Palacio   
543                                             Wonder  R. J. Palacio   
544                                             Wonder  R. J. Palacio   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)    Jeff Kinney   
546  You Are a Badass: How to Stop Doubting Your Gr...    Jen Sincero   
547  You Are a Badass: How to Stop Doubting Your Gr...    Jen Sincero   
548  You Are a Badass: How to Stop Doubting Your Gr...    Jen Sincero   
549  You Are a Badass: How to Stop Doubting Your Gr...    Jen Sincero   

     User Rating  Reviews  Price  Year        Genre  
540          4.8    21625      9  2013      Fiction  
541          4.8    21625      9  2014      Fiction  
542          4.8    21625      9  2015      Fiction  
543          4.8    21625      9  2016      Fiction  
544          4.8    21625      9  2017      Fiction  
545          4.9     9413      8  2019      Fiction  
546          4.7    14331      8  2016  Non Fiction  
547          4.7    14331      8  2017  Non Fiction  
548          4.7    14331      8  2018  Non Fiction  
549          4.7    14331      8  2019  Non Fiction

In [53]:
"""Books are sorted by book title in alphabetical order."""

'Books are sorted by book title in alphabetical order.'

In [54]:
#4 Discover df objects with df.info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


In [95]:
libpandas.categorize_columns(df)

Continuous columns = ['User Rating', 'Reviews', 'Price', 'Year']
Categorical columns = ['Name', 'Author', 'Genre']


(['User Rating', 'Reviews', 'Price', 'Year'], ['Name', 'Author', 'Genre'])

In [55]:
#4-1 =  Range index and total entries.
range_index = df.index
range_index

RangeIndex(start=0, stop=550, step=1)

In [56]:
#4-2 = total columns
number_of_columns = len(df.columns)
number_of_columns

7

In [57]:
#4-3 non-null compared to total entries
df.isnull().sum()

Name           0
Author         0
User Rating    0
Reviews        0
Price          0
Year           0
Genre          0
dtype: int64

In [58]:
# Check for empty strings in the entire DataFrame
empty_strings_in_df = (df == '')
# This will print the total number of empty strings in the DataFrame
empty_strings_in_df.sum().sum()

np.int64(0)

""""
The result shows zeros for all columns, that means there are no missing values (NaN or None) in the DataFrame.
"""


In [59]:
#4-4 = DType and possible savings plans
type(df.dtypes)

pandas.core.series.Series

In [60]:
df.dtypes

Name            object
Author          object
User Rating    float64
Reviews          int64
Price            int64
Year             int64
Genre           object
dtype: object

In [61]:
# 4-5 : Memory usage.
tips_df_memory = df.memory_usage()
tips_df_memory

Index           132
Name           4400
Author         4400
User Rating    4400
Reviews        4400
Price          4400
Year           4400
Genre          4400
dtype: int64

In [62]:
#5 Describe
df.describe()

User Rating       Reviews       Price         Year
count   550.000000    550.000000  550.000000   550.000000
mean      4.618364  11953.281818   13.100000  2014.000000
std       0.226980  11731.132017   10.842262     3.165156
min       3.300000     37.000000    0.000000  2009.000000
25%       4.500000   4058.000000    7.000000  2011.000000
50%       4.700000   8580.000000   11.000000  2014.000000
75%       4.800000  17253.250000   16.000000  2017.000000
max       4.900000  87841.000000  105.000000  2019.000000

In [63]:
df_describe = df.describe()
df_describe

User Rating       Reviews       Price         Year
count   550.000000    550.000000  550.000000   550.000000
mean      4.618364  11953.281818   13.100000  2014.000000
std       0.226980  11731.132017   10.842262     3.165156
min       3.300000     37.000000    0.000000  2009.000000
25%       4.500000   4058.000000    7.000000  2011.000000
50%       4.700000   8580.000000   11.000000  2014.000000
75%       4.800000  17253.250000   16.000000  2017.000000
max       4.900000  87841.000000  105.000000  2019.000000

In [70]:
#5-1 : Mean average value.
mpg_mean_values = np.round(df_describe.loc["mean"],2)
mpg_mean_values

User Rating        4.62
Reviews        11953.28
Price             13.10
Year            2014.00
Name: mean, dtype: float64

In [71]:
#5-2 : std 

mpg_std_values = np.round(df_describe.loc["std"],2)
mpg_std_values

User Rating        0.23
Reviews        11731.13
Price             10.84
Year               3.17
Name: std, dtype: float64

In [72]:
#5-3 : min

mpg_min_values = np.round(df_describe.loc["min"])
mpg_min_values

User Rating       3.0
Reviews          37.0
Price             0.0
Year           2009.0
Name: min, dtype: float64

In [73]:
#5-3 : max

mpg_max_values = np.round(df_describe.loc["max"])
mpg_max_values

User Rating        5.0
Reviews        87841.0
Price            105.0
Year            2019.0
Name: max, dtype: float64

In [74]:
#5-4 : Quartiles 25% , 50% , 75% obs: 50% is median

mpg_quartiles = df_describe.loc[['25%', '50%', '75%']]
mpg_quartiles = np.round(mpg_quartiles, 2)
mpg_quartiles

User Rating   Reviews  Price    Year
25%          4.5   4058.00    7.0  2011.0
50%          4.7   8580.00   11.0  2014.0
75%          4.8  17253.25   16.0  2017.0

In [ ]:
"""
6- Retrieve data from data frame for 4-4 , and 
    especially Object then float64 then int64
    Check continues data and categorical data.
"""

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


In [77]:
df['Genre'].value_counts()

Genre
Non Fiction    310
Fiction        240
Name: count, dtype: int64

## Part 1
* A:    Use pandas to read the `bestsellers` dataset into a DataFrame 
* B:    Once you've done that, use Pandas to figure out how many rows and columns the DF has
* C:    Inspect the first 5 rows
* D:    Inspect the first 19 rows
* E:    Inspect the last 5 rows
* F:    Inspect the last 2 rows 
* G:    Which columns (if any) are missing values?
* H:    What datatype did Pandas assign to "User Rating"?
* I:    How many integer columns are in the DataFrame?

In [64]:
#A

df = pd.read_csv("bestsellers.csv")
df

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

In [65]:
#B
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


Rows = 549
Columns = 7

In [66]:
df.tail(20)


Name  \
530               What to Expect When You're Expecting   
531  Wheat Belly: Lose the Wheat, Lose the Weight, ...   
532  Wheat Belly: Lose the Wheat, Lose the Weight, ...   
533                            When Breath Becomes Air   
534                            Where the Crawdads Sing   
535                          Where the Wild Things Are   
536  Whose Boat Is This Boat?: Comments That Don't ...   
537  Wild: From Lost to Found on the Pacific Crest ...   
538  Winter of the World: Book Two of the Century T...   
539  Women Food and God: An Unexpected Path to Almo...   
540                                             Wonder   
541                                             Wonder   
542                                             Wonder   
543                                             Wonder   
544                                             Wonder   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                               Author  User Rating  Reviews  Price  Year  \
530                     Heidi Murkoff          4.4     3341      9  2011   
531                     William Davis          4.4     7497      6  2012   
532                     William Davis          4.4     7497      6  2013   
533                    Paul Kalanithi          4.8    13779     14  2016   
534                       Delia Owens          4.8    87841     15  2019   
535                    Maurice Sendak          4.8     9967     13  2009   
536  The Staff of The Late Show with
          4.6     6669     12  2018   
537                    Cheryl Strayed          4.4    17044     18  2012   
538                       Ken Follett          4.5    10760     15  2012   
539                       Geneen Roth          4.2     1302     11  2010   
540                     R. J. Palacio          4.8    21625      9  2013   
541                     R. J. Palacio          4.8    21625      9  2014   
542                     R. J. Palacio          4.8    21625      9  2015   
543                     R. J. Palacio          4.8    21625      9  2016   
544                     R. J. Palacio          4.8    21625      9  2017   
545                       Jeff Kinney          4.9     9413      8  2019   
546                       Jen Sincero          4.7    14331      8  2016   
547                       Jen Sincero          4.7    14331      8  2017   
548                       Jen Sincero          4.7    14331      8  2018   
549                       Jen Sincero          4.7    14331      8  2019   

           Genre  
530  Non Fiction  
531  Non Fiction  
532  Non Fiction  
533  Non Fiction  
534      Fiction  
535      Fiction  
536  Non Fiction  
537  Non Fiction  
538      Fiction  
539  Non Fiction  
540      Fiction  
541      Fiction  
542      Fiction  
543      Fiction  
544      Fiction  
545      Fiction  
546  Non Fiction  
547  Non Fiction  
548  Non Fiction  
549  Non Fiction

In [67]:
duplicates = df[df["Name"].duplicated(keep=False)]
print("Rows with the same value:")
duplicates

Rows with the same value:


Name           Author  \
9                            A Man Called Ove: A Novel  Fredrik Backman   
10                           A Man Called Ove: A Novel  Fredrik Backman   
20                         All the Light We Cannot See    Anthony Doerr   
21                         All the Light We Cannot See    Anthony Doerr   
32                                            Becoming   Michelle Obama   
..                                                 ...              ...   
544                                             Wonder    R. J. Palacio   
546  You Are a Badass: How to Stop Doubting Your Gr...      Jen Sincero   
547  You Are a Badass: How to Stop Doubting Your Gr...      Jen Sincero   
548  You Are a Badass: How to Stop Doubting Your Gr...      Jen Sincero   
549  You Are a Badass: How to Stop Doubting Your Gr...      Jen Sincero   

     User Rating  Reviews  Price  Year        Genre  
9            4.6    23848      8  2016      Fiction  
10           4.6    23848      8  2017      Fiction  
20           4.6    36348     14  2014      Fiction  
21           4.6    36348     14  2015      Fiction  
32           4.8    61133     11  2018  Non Fiction  
..           ...      ...    ...   ...          ...  
544          4.8    21625      9  2017      Fiction  
546          4.7    14331      8  2016  Non Fiction  
547          4.7    14331      8  2017  Non Fiction  
548          4.7    14331      8  2018  Non Fiction  
549          4.7    14331      8  2019  Non Fiction  

[295 rows x 7 columns]

In [84]:
merged = df.groupby("Name").agg({
    "Year": lambda x: ', '.join(map(str, sorted(x)))
}).reset_index()

print(merged)

                                                  Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
346  Winter of the World: Book Two of the Century T...   
347  Women Food and God: An Unexpected Path to Almo...   
348                                             Wonder   
349       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
350  You Are a Badass: How to Stop Doubting Your Gr...   

                             Year  
0                            2016  
1                            2011  
2                            2018  
3                            2017  
4                            2019  
..                            ...  
346                          2012  
347            

In [93]:
is_unique_col1 = merged["Name"].is_unique
is_unique_col1

KeyError: 'Name'

In [92]:
merged

Author  \
Name                                                                           
10-Day Green Smoothie Cleanse                                       JJ Smith   
11/22/63: A Novel                                               Stephen King   
12 Rules for Life: An Antidote to Chaos                   Jordan B. Peterson   
1984 (Signet Classics)                                         George Orwell   
5,000 Awesome Facts (About Everything!) (Nation...  National Geographic Kids   
...                                                                      ...   
Wrecking Ball (Diary of a Wimpy Kid Book 14)                     Jeff Kinney   
You Are a Badass: How to Stop Doubting Your Gre...               Jen Sincero   
You Are a Badass: How to Stop Doubting Your Gre...               Jen Sincero   
You Are a Badass: How to Stop Doubting Your Gre...               Jen Sincero   
You Are a Badass: How to Stop Doubting Your Gre...               Jen Sincero   

                                                    User Rating  Reviews  \
Name                                                                       
10-Day Green Smoothie Cleanse                               4.7    17350   
11/22/63: A Novel                                           4.6     2052   
12 Rules for Life: An Antidote to Chaos                     4.7    18979   
1984 (Signet Classics)                                      4.7    21424   
5,000 Awesome Facts (About Everything!) (Nation...          4.8     7665   
...                                                         ...      ...   
Wrecking Ball (Diary of a Wimpy Kid Book 14)                4.9     9413   
You Are a Badass: How to Stop Doubting Your Gre...          4.7    14331   
You Are a Badass: How to Stop Doubting Your Gre...          4.7    14331   
You Are a Badass: How to Stop Doubting Your Gre...          4.7    14331   
You Are a Badass: How to Stop Doubting Your Gre...          4.7    14331   

                                                    Price  Year        Genre  
Name                                                                          
10-Day Green Smoothie Cleanse                           8  2016  Non Fiction  
11/22/63: A Novel                                      22  2011      Fiction  
12 Rules for Life: An Antidote to Chaos                15  2018  Non Fiction  
1984 (Signet Classics)                                  6  2017      Fiction  
5,000 Awesome Facts (About Everything!) (Nation...     12  2019  Non Fiction  
...                                                   ...   ...          ...  
Wrecking Ball (Diary of a Wimpy Kid Book 14)            8  2019      Fiction  
You Are a Badass: How to Stop Doubting Your Gre...      8  2016  Non Fiction  
You Are a Badass: How to Stop Doubting Your Gre...      8  2017  Non Fiction  
You Are a Badass: How to Stop Doubting Your Gre...      8  2018  Non Fiction  
You Are a Badass: How to Stop Doubting Your Gre...      8  2019  Non Fiction  

[550 rows x 6 columns]

In [91]:
new_df = merged.set_index("Name")
new_df.tail(20)

KeyError: "None of ['Name'] are in the columns"

## Part 2
* A:    Find the lowest User Rating in the DF
* B:    Find the highest Price in the DF
* C:    What is the average User Rating?
* D:    What the average User Rating of the first 5 rows?
* E:    What User Review score appeared the most?
* F:    What is the total (sum) of all the values in the Reviews column?
* G:    How many different authors are featured in the dataset?
* H:    Which author wrote the most number of books on the list?  How many did they write?

## Part 3
* A:    Retrieve a series that contains the book Names
* B:    Retrieve a series that contains the User Ratings
* C:    Retrieve the first 8 Authors
* D:    Find the unique Genres
* E:    Find the number of unique Authors
* F:    Find the average Price
* G:    Find the 10 highest prices
* H:    Find the top 3 most common book titles in the dataset
* I:    Create a new dataframe with only Author and User Rating.
* J:    Using the new dataframe, find the most common combination of Author and User Rating Score.

## Part 4
* A:    Find the books written by Pete Souza
* B:    Find the books that are under 10 dollars Ratings
* C:    Find the books that have a price between 50 and 60 dollars
* D:    Find all the books written by Kristin Hannah, Andy Weir, or Delia Owens
* E:    Find the Non Fiction books that are rated 4.9
* F:    Find the fiction book with the lowest User Rating
* G:    Find 2012's top 5 Fiction books with the most Reviews